File: frontend/src/lib/schemas.ts
/** Zod schemas mirroring the backend Pydantic ConfigSchema. */

import { z } from "zod";

/** Normalise JSON null → undefined so .optional() fields don't receive null from the backend. */
const nullishToUndef = <T extends z.ZodTypeAny>(schema: T) =>
  z.preprocess((v) => v ?? undefined, schema);

export const OPERATORS = [">=", "<=", "==", "!=", "contains", "regex", "in"] as const;
export const operatorSchema = z.enum(OPERATORS);
export type Operator = z.infer<typeof operatorSchema>;

export const SOURCE_ROLES = ["primary", "lookup"] as const;
export const sourceRoleSchema = z.enum(SOURCE_ROLES);
export type SourceRole = z.infer<typeof sourceRoleSchema>;

export const JOIN_TYPES = ["left", "inner"] as const;
export const joinTypeSchema = z.enum(JOIN_TYPES);
export type JoinType = z.infer<typeof joinTypeSchema>;

/** A scalar Excel cell value. Conditions may also compare against a list (for `in`). */
export const cellValueSchema = z.union([z.string(), z.number(), z.boolean(), z.null()]);
export type CellValue = z.infer<typeof cellValueSchema>;

const qualified = (v: string) =>
  v.includes(".") && !v.startsWith(".") && !v.endsWith(".");

export const targetTemplateSchema = z.object({
  sheet: z.string().min(1),
  header_row: z.number().int().min(1),
  preserve_styles: z.boolean().default(true),
  columns: z.array(z.string()).min(1),
  sample_filename: nullishToUndef(z.string().optional()),
});

export const sourceSpecSchema = z.object({
  alias: z.string().min(1),
  role: sourceRoleSchema,
  sheet: z.string().min(1),
  header_row: z.number().int().min(1),
  sample_filename: nullishToUndef(z.string().optional()),
});

export const joinRuleSchema = z.object({
  left: z.string().refine(qualified, "err.qualified"),
  right: z.string().refine(qualified, "err.qualified"),
  type: joinTypeSchema.default("left"),
});

export const conditionSchema = z.object({
  field: z.string().refine(qualified, "err.qualified"),
  op: operatorSchema,
  value: z.union([cellValueSchema, z.array(cellValueSchema)]),
});

export const cellAddressPattern = /^[A-Z]+[1-9]\d*$/;

export const sourceCellSchema = z.object({
  alias: z.string().min(1),
  address: z.string().regex(cellAddressPattern, "err.cellAddress"),
});

export const mappingSchema = z
  .object({
    target: z.string().min(1),
    source: nullishToUndef(z.string().refine(qualified, "err.qualified").optional()),
    literal: cellValueSchema.optional(),
    source_cell: sourceCellSchema.nullable().optional(),
    conditions: z.array(conditionSchema).default([]),
    default: cellValueSchema.default(""),
  })
  .superRefine((m, ctx) => {
    const hasSource = m.source !== undefined && m.source !== "";
    const hasLiteral = m.literal !== undefined && m.literal !== null && m.literal !== "";
    const hasCell = m.source_cell !== undefined && m.source_cell !== null;
    const setCount = [hasSource, hasLiteral, hasCell].filter(Boolean).length;
    if (setCount > 1) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "err.xorMode",
        path: ["source"],
      });
    }
    if (setCount === 0) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "err.requireOneMode",
        path: ["source"],
      });
    }
  });

export const NAME_PATTERN = /^[\p{L}\p{N}_\- ]{1,80}$/u;

export const configSchema = z
  .object({
    version: z.string().default("1.0"),
    name: z
      .string()
      .transform((v) => v.trim())
      .refine((v) => NAME_PATTERN.test(v), "err.invalidName"),
    target_template: targetTemplateSchema,
    sources: z.array(sourceSpecSchema).min(1),
    joins: z.array(joinRuleSchema).default([]),
    mappings: z.array(mappingSchema).min(1),
  })
  .superRefine((cfg, ctx) => {
    const aliases = new Set(cfg.sources.map((s) => s.alias));
    if (aliases.size !== cfg.sources.length) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: "err.duplicateAlias", path: ["sources"] });
    }
    if (!cfg.sources.some((s) => s.role === "primary")) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: "err.noPrimary", path: ["sources"] });
    }
    cfg.joins.forEach((j, i) => {
      for (const side of [j.left, j.right]) {
        const a = side.split(".", 1)[0];
        if (!aliases.has(a)) {
          ctx.addIssue({ code: z.ZodIssueCode.custom, message: "err.joinUnknownAlias", params: { alias: a }, path: ["joins", i] });
        }
      }
    });
    cfg.mappings.forEach((m, i) => {
      if (m.source) {
        const a = m.source.split(".", 1)[0];
        if (!aliases.has(a)) {
          ctx.addIssue({ code: z.ZodIssueCode.custom, message: "err.mappingSourceUnknownAlias", params: { alias: a }, path: ["mappings", i, "source"] });
        }
      }
      if (m.source_cell && !aliases.has(m.source_cell.alias)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "err.mappingCellUnknownAlias",
          params: { alias: m.source_cell.alias },
          path: ["mappings", i, "source_cell", "alias"],
        });
      }
    });
  });

export type Condition = z.infer<typeof conditionSchema>;
export type Mapping = z.infer<typeof mappingSchema>;
export type SourceSpec = z.infer<typeof sourceSpecSchema>;
export type JoinRule = z.infer<typeof joinRuleSchema>;
export type TargetTemplate = z.infer<typeof targetTemplateSchema>;
export type Config = z.infer<typeof configSchema>;

// ---------- Runtime DTOs ----------

export type JobStatus = "pending" | "running" | "done" | "failed" | "cancelled";
export type SubtaskStatus = "pending" | "running" | "done" | "failed";

export type JobSnapshot = {
  job_id: string;
  status: JobStatus;
  total: number;
  done: number;
  failed: number;
  eta_seconds?: number | null;
  error?: string | null;
  config_name?: string | null;
  /** ISO 8601 expiry time for re-download; absent before first download or after job dir is purged. */
  download_expires_at?: string | null;
};

export type SubtaskState = {
  source_file: string;
  status: SubtaskStatus;
  duration_ms?: number | null;
  user_message?: string | null;
  tech_detail?: string | null;
};

export type JobState = {
  job_id: string;
  config_name?: string | null;
  status: JobStatus;
  created_at: string;
  download_started_at?: string | null;
  subtasks: Record<string, SubtaskState>;
  cancel_requested: boolean;
};

File: backend/app/schemas.py
"""Pydantic schemas: ConfigSchema (the json a user authors in ConfigBuilder)
plus runtime DTOs used by the API and worker.
"""

from __future__ import annotations

import re
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

NAME_PATTERN = re.compile(r"^[\w一-鿿぀-ヿ\- ]{1,80}$", re.UNICODE)

Operator = Literal[">=", "<=", "==", "!=", "contains", "regex", "in"]
SourceRole = Literal["primary", "lookup"]
JoinType = Literal["left", "inner", "outer", "right"]
JobStatus = Literal["pending", "running", "done", "failed", "cancelled"]
SubtaskStatus = Literal["pending", "running", "done", "failed"]


class TargetTemplate(BaseModel):
    sheet: str
    header_row: int = Field(ge=1)
    preserve_styles: bool = True
    columns: list[str] = Field(min_length=1)
    # Filename used when the config was last saved — shown as a hint in BatchRunner.
    sample_filename: str | None = None


class SourceSpec(BaseModel):
    alias: str = Field(min_length=1)
    role: SourceRole
    sheet: str
    header_row: int = Field(ge=1)
    sample_filename: str | None = None


class JoinRule(BaseModel):
    left: str
    right: str
    type: JoinType = "left"

    @field_validator("left", "right")
    @classmethod
    def _qualified(cls, v: str) -> str:
        if "." not in v or v.startswith(".") or v.endswith("."):
            raise ValueError(f"必須為 alias.column 格式，收到 {v!r}")
        return v


class Condition(BaseModel):
    field: str
    op: Operator
    value: Any

    @field_validator("field")
    @classmethod
    def _qualified(cls, v: str) -> str:
        if "." not in v:
            raise ValueError(f"條件欄位必須為 alias.column 格式，收到 {v!r}")
        return v


class SourceCell(BaseModel):
    """Absolute cell reference into a source's xlsx sheet, bypassing the header
    abstraction. The sheet is the one defined on the matching SourceSpec.
    """

    alias: str = Field(min_length=1)
    address: str = Field(pattern=r"^[A-Z]+[1-9]\d*$")


class Mapping(BaseModel):
    target: str = Field(min_length=1)
    source: str | None = None
    literal: Any = None
    source_cell: SourceCell | None = None
    conditions: list[Condition] = Field(default_factory=list)
    default: Any = ""

    @field_validator("source")
    @classmethod
    def _qualified(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if "." not in v:
            raise ValueError(f"映射來源必須為 alias.column 格式，收到 {v!r}")
        return v

    @model_validator(mode="after")
    def _exactly_one_value_source(self) -> "Mapping":
        set_count = sum(
            x is not None for x in (self.source, self.literal, self.source_cell)
        )
        if set_count == 0:
            raise ValueError(
                f"mapping {self.target!r} 必須指定 source、source_cell 或 literal 其中一個"
            )
        if set_count > 1:
            raise ValueError(
                f"mapping {self.target!r} 的 source / source_cell / literal 只能擇一"
            )
        return self


class ConfigSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version: str = "1.0"
    name: str
    target_template: TargetTemplate
    sources: list[SourceSpec] = Field(min_length=1)
    joins: list[JoinRule] = Field(default_factory=list)
    mappings: list[Mapping] = Field(min_length=1)

    @field_validator("name")
    @classmethod
    def _name(cls, v: str) -> str:
        v = v.strip()
        if not NAME_PATTERN.match(v):
            raise ValueError("專案名稱僅可含中英文 / 數字 / 底線 / 連字號 / 空格，長度 1–80")
        return v

    @model_validator(mode="after")
    def _validate_relationships(self) -> "ConfigSchema":
        aliases = {s.alias for s in self.sources}
        if len({s.alias for s in self.sources}) != len(self.sources):
            raise ValueError("sources 中 alias 不可重複")
        primaries = [s for s in self.sources if s.role == "primary"]
        if not primaries:
            raise ValueError("至少需要一個 role=primary 的 source")

        for j in self.joins:
            for side in (j.left, j.right):
                alias = side.split(".", 1)[0]
                if alias not in aliases:
                    raise ValueError(f"join 引用了不存在的 source alias: {alias}")

        for m in self.mappings:
            if m.source is not None:
                alias = m.source.split(".", 1)[0]
                if alias not in aliases:
                    raise ValueError(f"mapping.source 引用了不存在的 alias: {alias}")
            if m.source_cell is not None and m.source_cell.alias not in aliases:
                raise ValueError(
                    f"mapping.source_cell 引用了不存在的 alias: {m.source_cell.alias}"
                )
            for c in m.conditions:
                a = c.field.split(".", 1)[0]
                if a not in aliases:
                    raise ValueError(f"condition.field 引用了不存在的 alias: {a}")

        return self

    @property
    def primary_alias(self) -> str:
        return next(s.alias for s in self.sources if s.role == "primary")

    @property
    def lookup_aliases(self) -> list[str]:
        return [s.alias for s in self.sources if s.role == "lookup"]


# ---------- Runtime DTOs ----------

class SubtaskState(BaseModel):
    source_file: str
    status: SubtaskStatus = "pending"
    duration_ms: int | None = None
    user_message: str | None = None
    tech_detail: str | None = None


class JobState(BaseModel):
    """Persisted in /data/jobs/{id}/state.json AND mirrored to Redis."""

    job_id: str
    config_name: str | None = None
    status: JobStatus = "pending"
    created_at: str
    download_started_at: str | None = None
    subtasks: dict[str, SubtaskState] = Field(default_factory=dict)
    cancel_requested: bool = False

    @property
    def total(self) -> int:
        return len(self.subtasks)

    @property
    def done(self) -> int:
        return sum(1 for s in self.subtasks.values() if s.status == "done")

    @property
    def failed(self) -> int:
        return sum(1 for s in self.subtasks.values() if s.status == "failed")


class JobSnapshot(BaseModel):
    """Public-facing summary returned by `GET /api/jobs/{id}`."""

    job_id: str
    status: JobStatus
    total: int
    done: int
    failed: int
    eta_seconds: int | None = None
    error: str | None = None
    config_name: str | None = None
